File: backend-ts/templates/selenium/java/pom/base/src/main/java/__BASE_PACKAGE_PATH__/base/BasePage.java
package {{BASE_PACKAGE}}.base;

import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Duration;

public abstract class BasePage {

    protected final WebDriver driver;
    protected final WebDriverWait wait;
    protected final Logger log = LoggerFactory.getLogger(getClass());

    protected BasePage(WebDriver driver) {
        this.driver = driver;
        this.wait   = new WebDriverWait(driver, Duration.ofSeconds({{TIMEOUT}} / 1000));
    }

    protected WebElement find(By locator) {
        return wait.until(ExpectedConditions.visibilityOfElementLocated(locator));
    }

    protected void click(By locator) {
        find(locator).click();
    }

    protected void type(By locator, String text) {
        WebElement el = find(locator);
        el.clear();
        el.sendKeys(text);
    }

    protected String getText(By locator) {
        return find(locator).getText();
    }

    protected boolean isVisible(By locator) {
        try { return find(locator).isDisplayed(); }
        catch (Exception e) { return false; }
    }

    protected void scrollTo(By locator) {
        WebElement el = driver.findElement(locator);
        ((JavascriptExecutor) driver).executeScript("arguments[0].scrollIntoView(true);", el);
    }

    protected void navigateTo(String path) {
        driver.get("{{BASE_URL}}" + path);
    }

    public String getTitle() {
        return driver.getTitle();
    }
}

File: backend-ts/templates/cypress/ts/page-objects/base/cypress/pages/BasePage.ts
export abstract class BasePage {
  protected readonly baseUrl = Cypress.config('baseUrl') ?? '';

  visit(path = '') {
    cy.visit(path);
    return this;
  }

  getTitle() {
    return cy.title();
  }

  protected find(selector: string) {
    return cy.get(selector);
  }

  protected click(selector: string) {
    cy.get(selector).click();
    return this;
  }

  protected type(selector: string, text: string) {
    cy.get(selector).clear().type(text);
    return this;
  }

  protected assertVisible(selector: string) {
    cy.get(selector).should('be.visible');
    return this;
  }

  protected assertText(selector: string, expected: string) {
    cy.get(selector).should('have.text', expected);
    return this;
  }

  protected waitForNetworkIdle() {
    cy.intercept('**').as('anyRequest');
    cy.wait('@anyRequest');
    return this;
  }
}
